### ISSUE TITLE
multistream not consuming entire stream due to race condition. reproduction included.

### ISSUE THREAD

**Issue by guanzo** [feross/multistream#50]
I'm running into an issue where multistream doesn't consume the entire stream. Not sure why. If you follow the instructions and run `node multistream.js` lots of times, you'll see intermittent content length mismatches. 

I included `node passthrough.js` to provide an example that never fails. `multistream.js` and `passthrough.js` are nearly identical, only difference being piping to multistream vs piping to a passthrough stream.

https://www.diffchecker.com/bi5wMRCW

Reproduction steps
```shell
$ git clone https://github.com/guanzo/multistream-debug
$ cd multistream-debug
$ npm i 

# Run 20+ times, no failures
$ node passthrough.js

# Run 20+ times, intermittent failures
$ node multistream.js
```

here's what i'm seeing
![](https://i.gyazo.com/5c46d81da53a0390ef99fc69d1fce5a4.png)

Edit: 
I found this bug after seeing webtorrent make never ending requests due to an invalid torrentFile. `create-torrent` uses `multistream`, and whenever this bug occurs, an invalid torrentFile is returned. Webseed tries to verify a valid chunk against an invalid torrentFile's chunk hash. The verification fails, so webtorrent retries endlessly.

Edit:
`$ node multistream.js` works with `multistream@3.0.0`. This commit seems to be the problem.

https://github.com/feross/multistream/commit/0813df4cc0227b61ad24f314d4970b17d8b45941

Removing `&& this._drained` fixes the bug. Thoughts?

**Comment by feross** [comment]
I don't have time at the moment to look at this. I wonder if @mafintosh has any thoughts about this.

**Comment by mafintosh** [comment]
I’m also strapped for time, so not sure I can be of much help atm. @feross you might just want to revert that commit unless it’s fixing another issue I don’t know?

**Comment by murphyproject** [comment]
Hi, sorry to intervene. I was evaluating the package. 

`
 // From:

  while ((chunk = this._current.read()) !== null && this._drained) {
    this._drained = this.push(chunk)
  }

 // To:

  while (this._drained && (chunk = this._current.read()) !== null) {
    this._drained = this.push(chunk)
  }

`

can we check this._drained first? Because read() can return a chuck, but this._drained can be false. this chunk of data will get lost.

**Comment by johnsusek** [comment]
Can confirm #51 fixes this issue for me

**Comment by srolel** [comment]
I'm also seeing this, is the fix going to be merged?

**Comment by wombat9000** [comment]
+1

**Comment by feross** [comment]
Fixed by #51 and released as 4.0.1
